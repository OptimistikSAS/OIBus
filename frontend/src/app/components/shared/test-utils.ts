import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };
