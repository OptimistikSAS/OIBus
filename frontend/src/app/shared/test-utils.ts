import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';
import { Page } from './types';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };

export function toPage<T>(content: Array<T>, totalElements: number = content.length, num = 0, size = Math.max(20, totalElements)): Page<T> {
  return {
    content,
    totalElements,
    number: num,
    size,
    totalPages: Math.ceil(totalElements / size)
  };
}

export function emptyPage<T>(size = 20): Page<T> {
  return {
    totalElements: 0,
    content: [] as Array<T>,
    size,
    totalPages: 0,
    number: 0
  };
}
