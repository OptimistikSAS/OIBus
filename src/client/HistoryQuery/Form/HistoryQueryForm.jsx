import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import { Form, Row, Col, Label, Button, Container } from 'reactstrap'
import { useNavigate } from 'react-router-dom'
import { FaPauseCircle, FaPlayCircle, FaArrowLeft } from 'react-icons/fa'
import { act } from 'react-dom/test-utils'
import {
  OIbTitle,
  OIbCheckBox,
  OIbText,
  OIbTextArea,
} from '../../components/OIbForm'
import OIbDate from '../../components/OIbForm/OIbDate.jsx'
import { ConfigContext } from '../../context/ConfigContext.jsx'
import PointsSection from './PointsSection.jsx'
import apis from '../../services/apis'
import utils from '../../helpers/utils'
import { AlertContext } from '../../context/AlertContext.jsx'

const HistoryQueryForm = ({ query }) => {
  const [queryToUpdate, setQueryToUpdate] = useState(query)
  const { name, paused } = queryToUpdate
  const { newConfig } = React.useContext(ConfigContext)
  const [lastCompleted, setLastCompleted] = useState()
  const [errors, setErrors] = useState({})
  const dataSource = newConfig?.south?.dataSources.find(
    (southHandler) => southHandler.id === queryToUpdate.southId,
  )
  const application = newConfig?.north?.applications.find(
    (northHandler) => northHandler.id === queryToUpdate.northId,
  )

  const { setAlert } = React.useContext(AlertContext)
  const navigate = useNavigate()

  const handlePause = async () => {
    await apis.pauseHistoryQuery(queryToUpdate.id, { paused: !paused })
    setQueryToUpdate({ ...queryToUpdate, paused: !queryToUpdate.paused })
  }

  const handleEnabled = async () => {
    await apis.enableHistoryQuery(queryToUpdate.id, { enabled: !queryToUpdate.enabled })
    setQueryToUpdate({ ...queryToUpdate, enabled: !queryToUpdate.enabled })
  }

  useEffect(() => {
    apis
      .getLastCompletedForHistoryQuery(queryToUpdate.id)
      .then((response) => {
        act(() => {
          setLastCompleted(response)
        })
      })
      .catch((error) => {
        console.error(error)
      })
  }, [])

  const handleUpdateHistoryQuery = async () => {
    if (Object.values(errors).length === 0) {
      setAlert(null)
      await apis.updateHistoryQuery(queryToUpdate.id, { ...queryToUpdate })
    } else {
      setAlert({ text: 'You have unresolved errors in configuration! Please fix before updating the settings.', type: 'danger' })
    }
  }

  const onChange = (propertyName, value) => {
    setQueryToUpdate({ ...queryToUpdate, [propertyName]: value })
  }

  const handleAddQuery = (_, value) => {
    setQueryToUpdate({ ...queryToUpdate, settings: { query: value } })
  }

  const handleAddPoint = (attributes) => {
    const newPointAttributes = {}
    attributes.forEach((attribute) => { newPointAttributes[attribute] = '' })
    setQueryToUpdate(
      { ...queryToUpdate, settings: { points: [...queryToUpdate.settings.points, { ...newPointAttributes }] } },
    )
  }

  const handleChangePoint = (attributeDescription, value, validity) => {
    if (validity) {
      setErrors((oldErrors) => ({ ...oldErrors, [attributeDescription]: validity }))
    } else {
      const copyErrorObject = errors
      delete copyErrorObject[attributeDescription]
      setErrors(copyErrorObject)
    }

    const attributeName = attributeDescription.split('.')[2]
    const pointIndex = Number(attributeDescription.split('.')[1])
    setQueryToUpdate((oldQuery) => ({
      ...oldQuery,
      settings: {
        points: oldQuery.settings.points.map((point, index) => {
          if (pointIndex === index) {
            return { ...point, [attributeName]: value }
          }
          return point
        }),
      },
    }))
  }

  const handleDeletePoint = (indexInTable) => {
    setQueryToUpdate((oldQuery) => ({
      ...oldQuery,
      settings: { points: oldQuery.settings.points.filter((_point, index) => index !== indexInTable) },
    }))
    Object.keys(errors).forEach((errorKey) => {
      if (errorKey.split('.')[1] === String(indexInTable)) delete errors[errorKey]
    })
  }

  const handleDeleteAllPoint = () => {
    setQueryToUpdate({ ...queryToUpdate, settings: { points: [] } })
  }

  const handleImportPoints = async (file) => {
    try {
      const text = await utils.readFileContent(file)
      utils
        .parseCSV(text)
        .then((newPoints) => {
          setQueryToUpdate({ ...queryToUpdate, settings: { points: [...queryToUpdate.settings.points, ...newPoints] } })
        })
        .catch((error) => {
          console.error(error)
          setAlert({ text: error.message, type: 'danger' })
        })
    } catch (error) {
      console.error(error)
      setAlert({ text: error.message, type: 'danger' })
    }
  }

  return (
    <>
      <div className="d-flex align-items-center w-100 oi-sub-nav mb-2">
        <h6 className="text-muted d-flex align-items-center pl-3 pt-1 ml-2">
          <Button
            id="oi-navigate"
            outline
            onClick={() => {
              navigate(-1)
            }}
            className="util-button"
          >
            <FaArrowLeft className="oi-back-icon mr-2" />
          </Button>
          {`| ${name}`}
          {paused ? (
            <>
              <FaPauseCircle
                className="oi-icon-breadcrumb"
                size={15}
                onClick={(e) => {
                  e.preventDefault()
                  handlePause()
                }}
              />
              <Label className="status-text-breadcrumb text-success">
                Ongoing
              </Label>
            </>
          ) : (
            <>
              <FaPlayCircle
                className="oi-icon-breadcrumb"
                size={15}
                onClick={(e) => {
                  e.preventDefault()
                  handlePause()
                }}
              />
              <Label className="status-text-breadcrumb text-warning">
                Paused
              </Label>
            </>
          )}
        </h6>
      </div>
      <Container fluid>
        <Form className="m-2">
          <OIbTitle label="Handlers" />
          <Row className="mb-2">
            <Col md={2}>
              <a href={`/south/${queryToUpdate.southId}`}>{dataSource?.name}</a>
              {'  ->  '}
              <a href={`/north/${queryToUpdate.northId}`}>{application?.name}</a>
            </Col>
          </Row>
          <OIbTitle label="General settings" />
          <Row>
            <Col md={2}>
              <OIbCheckBox
                name="enabled"
                label="Enabled"
                defaultValue={false}
                value={!!queryToUpdate.enabled}
                help={<div>Enable this history query</div>}
                onChange={handleEnabled}
                switchButton
              />
            </Col>
          </Row>
          <Row>
            <Col md={2}>
              <OIbDate
                name="startTime"
                value={queryToUpdate.startTime}
                label="From date"
                onChange={onChange}
              />
            </Col>
            <Col md={2}>
              <OIbDate
                name="endTime"
                value={queryToUpdate.endTime}
                label="To date"
                onChange={onChange}
              />
            </Col>
            <Col md={2}>
              <OIbText
                label="File pattern"
                name="filePattern"
                value={queryToUpdate.filePattern}
                defaultValue="./@ConnectorName-@CurrentDate-@QueryPart.csv"
                onChange={onChange}
              />
            </Col>
            <Col md={2}>
              <OIbCheckBox
                name="compress"
                label="Compress"
                defaultValue={false}
                value={!!queryToUpdate.compress}
                onChange={onChange}
              />
            </Col>
          </Row>
          <Row>
            <Col md={4}>
              <div>
                <strong>Last completed dates: </strong>
                <div className="mb-2">
                  {lastCompleted?.south.length ? (
                    <table className="last-completed-table">
                      <thead>
                        <tr>
                          <th>Scan mode</th>
                          <th>Last completed date</th>
                        </tr>
                      </thead>
                      <tbody>
                        {lastCompleted?.south.map((obj) => (
                          <tr>
                            <td>{obj.scanMode}</td>
                            <td>{obj.lastCompletedDate}</td>
                          </tr>
                        ))}
                      </tbody>
                    </table>
                  ) : (
                    'Nothing to show'
                  )}
                </div>
              </div>
            </Col>
          </Row>
          <Row>
            <Col md={8}>
              {queryToUpdate.settings.points ? (
                <PointsSection
                  query={queryToUpdate}
                  handleAddPoint={handleAddPoint}
                  handleChange={handleChangePoint}
                  handleDeletePoint={handleDeletePoint}
                  handleDeleteAllPoint={handleDeleteAllPoint}
                  handleImportPoints={handleImportPoints}
                />
              ) : (
                <>
                  <OIbTitle label="Request" />
                  <OIbTextArea
                    label="Query"
                    name="query"
                    value={queryToUpdate.settings?.query}
                    onChange={handleAddQuery}
                  />
                </>
              )}
            </Col>
          </Row>
          <Row>
            <Col md={2}>
              <Button color="primary" onClick={() => handleUpdateHistoryQuery()}>
                Save history query
              </Button>
            </Col>
          </Row>
        </Form>
      </Container>
    </>
  )
}

HistoryQueryForm.propTypes = { query: PropTypes.object.isRequired }

export default HistoryQueryForm
