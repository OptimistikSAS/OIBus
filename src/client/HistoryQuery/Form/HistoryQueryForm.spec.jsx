/**
 * @jest-environment jsdom
 */
import React from 'react'
import ReactDOM from 'react-dom'
import { act } from 'react-dom/test-utils'

import newConfig from '../../../../tests/testHistoryConfig'
import HistoryQueryForm from './HistoryQueryForm.jsx'

React.useContext = jest.fn().mockReturnValue({ newConfig })
jest.mock('react-router-dom', () => (
  { useNavigate: jest.fn().mockReturnValue({}) }
))
const mockMath = Object.create(global.Math)
mockMath.random = () => 1
global.Math = mockMath

let container
beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

describe('HistoryQueryForm', () => {
  newConfig.forEach((historyQuery) => {
    test(`check HistoryQueryForm with data: ${historyQuery.name}`, () => {
      act(() => {
        ReactDOM.render(
          <HistoryQueryForm
            query={historyQuery}
          />,
          container,
        )
      })
      expect(container).toMatchSnapshot()
    })
  })

  test('check HistoryQueryForm with empty application', () => {
    act(() => {
      ReactDOM.render(
        <HistoryQueryForm
          query={{
            id: 'unique-id',
            name: 'north -> south',
            enabled: false,
            status: 'pending',
            southId: 'southId',
            northId: 'northId',
            query: '',
            order: 1,
            settings: {},
          }}
        />,
        container,
      )
    })
    expect(container).toMatchSnapshot()
  })
})
