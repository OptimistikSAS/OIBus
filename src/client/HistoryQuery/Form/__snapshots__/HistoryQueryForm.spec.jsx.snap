// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryQueryForm check HistoryQueryForm with data: MSSQL South -> Console 1`] = `
<div>
  <div
    class="d-flex align-items-center w-100 oi-sub-nav mb-2"
  >
    <h6
      class="text-muted d-flex align-items-center pl-3 pt-1 ml-2"
    >
      <button
        class="util-button btn btn-outline-secondary"
        id="oi-navigate"
        type="button"
      >
        <svg
          class="oi-back-icon mr-2"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
          />
        </svg>
      </button>
      | MSSQL South -&gt; Console
      <svg
        class="oi-icon-breadcrumb"
        fill="currentColor"
        height="15"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm-16 328c0 8.8-7.2 16-16 16h-48c-8.8 0-16-7.2-16-16V176c0-8.8 7.2-16 16-16h48c8.8 0 16 7.2 16 16v160zm112 0c0 8.8-7.2 16-16 16h-48c-8.8 0-16-7.2-16-16V176c0-8.8 7.2-16 16-16h48c8.8 0 16 7.2 16 16v160z"
        />
      </svg>
      <label
        class="status-text-breadcrumb text-success form-label"
      >
        Ongoing
      </label>
    </h6>
  </div>
  <div
    class="container-fluid"
  >
    <form
      class="m-2"
    >
      <div
        class="row"
      >
        <h5>
          Handlers
        </h5>
      </div>
      <div
        class="container-fluid"
      />
      <div
        class="mb-2 row"
      >
        <div
          class="col-md-2"
        >
          <a
            href="/south/southId-2"
          />
            -&gt;  
          <a
            href="/north/northId-2"
          />
        </div>
      </div>
      <div
        class="row"
      >
        <h5>
          General settings
        </h5>
      </div>
      <div
        class="container-fluid"
      />
      <div
        class="row"
      >
        <div
          class="col-md-2"
        >
          <div
            class="form-check form-switch"
          >
            <input
              checked=""
              class="mb-3 form-check-input"
              color="secondary"
              id="enabled"
              label="Enabled"
              name="enabled"
              type="checkbox"
            />
            <small
              class="form-text text-muted"
            >
              <div>
                Enable this history query
              </div>
            </small>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="startTime"
            >
              From date
            </label>
            <input
              class="oi-form-input form-control"
              id="startTime"
              placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
              required=""
              type="datetime-local"
              value="2021-10-15T14:18"
            />
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="endTime"
            >
              To date
            </label>
            <input
              class="oi-form-input form-control"
              id="endTime"
              placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
              required=""
              type="datetime-local"
              value="2021-10-16T14:18"
            />
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="filePattern"
            >
              File pattern
            </label>
            <input
              aria-invalid="false"
              class="oi-form-input form-control"
              id="filePattern"
              name="filePattern"
              type="text"
              value="history1-@CurrentDate.csv"
            />
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="mb-3 mb-3"
          >
            <label
              class="form-check-label form-label"
            >
              Compress
            </label>
            <input
              checked=""
              class="oi-form-input checkbox form-check-input"
              id="compress"
              name="compress"
              type="checkbox"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-4"
        >
          <div>
            <strong>
              Last completed dates: 
            </strong>
            <div
              class="mb-2"
            >
              Nothing to show
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-8"
        >
          <div
            class="spinner-container"
          >
            <div
              class="spinner-grow text-primary"
              role="status"
            >
              <span
                class="visually-hidden"
              >
                Loading...
              </span>
            </div>
            ...loading points from OIBus server...
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-2"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Save history query
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`HistoryQueryForm check HistoryQueryForm with data: MSSQL South -> OIAnalytics North 1`] = `
<div>
  <div
    class="d-flex align-items-center w-100 oi-sub-nav mb-2"
  >
    <h6
      class="text-muted d-flex align-items-center pl-3 pt-1 ml-2"
    >
      <button
        class="util-button btn btn-outline-secondary"
        id="oi-navigate"
        type="button"
      >
        <svg
          class="oi-back-icon mr-2"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
          />
        </svg>
      </button>
      | MSSQL South -&gt; OIAnalytics North
      <svg
        class="oi-icon-breadcrumb"
        fill="currentColor"
        height="15"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm-16 328c0 8.8-7.2 16-16 16h-48c-8.8 0-16-7.2-16-16V176c0-8.8 7.2-16 16-16h48c8.8 0 16 7.2 16 16v160zm112 0c0 8.8-7.2 16-16 16h-48c-8.8 0-16-7.2-16-16V176c0-8.8 7.2-16 16-16h48c8.8 0 16 7.2 16 16v160z"
        />
      </svg>
      <label
        class="status-text-breadcrumb text-success form-label"
      >
        Ongoing
      </label>
    </h6>
  </div>
  <div
    class="container-fluid"
  >
    <form
      class="m-2"
    >
      <div
        class="row"
      >
        <h5>
          Handlers
        </h5>
      </div>
      <div
        class="container-fluid"
      />
      <div
        class="mb-2 row"
      >
        <div
          class="col-md-2"
        >
          <a
            href="/south/southId-1"
          />
            -&gt;  
          <a
            href="/north/northId-1"
          />
        </div>
      </div>
      <div
        class="row"
      >
        <h5>
          General settings
        </h5>
      </div>
      <div
        class="container-fluid"
      />
      <div
        class="row"
      >
        <div
          class="col-md-2"
        >
          <div
            class="form-check form-switch"
          >
            <input
              checked=""
              class="mb-3 form-check-input"
              color="secondary"
              id="enabled"
              label="Enabled"
              name="enabled"
              type="checkbox"
            />
            <small
              class="form-text text-muted"
            >
              <div>
                Enable this history query
              </div>
            </small>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="startTime"
            >
              From date
            </label>
            <input
              class="oi-form-input form-control"
              id="startTime"
              placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
              required=""
              type="datetime-local"
              value="2021-10-15T14:18"
            />
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="endTime"
            >
              To date
            </label>
            <input
              class="oi-form-input form-control"
              id="endTime"
              placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
              required=""
              type="datetime-local"
              value="2021-10-16T14:18"
            />
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="filePattern"
            >
              File pattern
            </label>
            <input
              aria-invalid="false"
              class="oi-form-input form-control"
              id="filePattern"
              name="filePattern"
              type="text"
              value="history1-@CurrentDate.csv"
            />
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="mb-3 mb-3"
          >
            <label
              class="form-check-label form-label"
            >
              Compress
            </label>
            <input
              checked=""
              class="oi-form-input checkbox form-check-input"
              id="compress"
              name="compress"
              type="checkbox"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-4"
        >
          <div>
            <strong>
              Last completed dates: 
            </strong>
            <div
              class="mb-2"
            >
              Nothing to show
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-8"
        >
          <div
            class="spinner-container"
          >
            <div
              class="spinner-grow text-primary"
              role="status"
            >
              <span
                class="visually-hidden"
              >
                Loading...
              </span>
            </div>
            ...loading points from OIBus server...
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-2"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Save history query
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`HistoryQueryForm check HistoryQueryForm with empty application 1`] = `
<div>
  <div
    class="d-flex align-items-center w-100 oi-sub-nav mb-2"
  >
    <h6
      class="text-muted d-flex align-items-center pl-3 pt-1 ml-2"
    >
      <button
        class="util-button btn btn-outline-secondary"
        id="oi-navigate"
        type="button"
      >
        <svg
          class="oi-back-icon mr-2"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
          />
        </svg>
      </button>
      | north -&gt; south
      <svg
        class="oi-icon-breadcrumb"
        fill="currentColor"
        height="15"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm115.7 272l-176 101c-15.8 8.8-35.7-2.5-35.7-21V152c0-18.4 19.8-29.8 35.7-21l176 107c16.4 9.2 16.4 32.9 0 42z"
        />
      </svg>
      <label
        class="status-text-breadcrumb text-warning form-label"
      >
        Paused
      </label>
    </h6>
  </div>
  <div
    class="container-fluid"
  >
    <form
      class="m-2"
    >
      <div
        class="row"
      >
        <h5>
          Handlers
        </h5>
      </div>
      <div
        class="container-fluid"
      />
      <div
        class="mb-2 row"
      >
        <div
          class="col-md-2"
        >
          <a
            href="/south/southId"
          />
            -&gt;  
          <a
            href="/north/northId"
          />
        </div>
      </div>
      <div
        class="row"
      >
        <h5>
          General settings
        </h5>
      </div>
      <div
        class="container-fluid"
      />
      <div
        class="row"
      >
        <div
          class="col-md-2"
        >
          <div
            class="form-check form-switch"
          >
            <input
              class="mb-3 form-check-input"
              color="secondary"
              id="enabled"
              label="Enabled"
              name="enabled"
              type="checkbox"
            />
            <small
              class="form-text text-muted"
            >
              <div>
                Enable this history query
              </div>
            </small>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="startTime"
            >
              From date
            </label>
            <input
              class="oi-form-input form-control"
              id="startTime"
              placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
              required=""
              type="datetime-local"
              value="1970-01-01T00:00"
            />
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
        <div
          class="col-md-2"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="endTime"
            >
              To date
            </label>
            <input
              class="oi-form-input form-control"
              id="endTime"
              placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
              required=""
              type="datetime-local"
              value="1970-01-01T00:00"
            />
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
        <div
          class="col-md-2"
        />
        <div
          class="col-md-2"
        >
          <div
            class="mb-3 mb-3"
          >
            <label
              class="form-check-label form-label"
            >
              Compress
            </label>
            <input
              class="oi-form-input checkbox form-check-input"
              id="compress"
              name="compress"
              type="checkbox"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-4"
        >
          <div>
            <strong>
              Last completed dates: 
            </strong>
            <div
              class="mb-2"
            >
              Nothing to show
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-8"
        >
          <div
            class="row"
          >
            <h5>
              Request
            </h5>
          </div>
          <div
            class="container-fluid"
          />
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="query"
            >
              Query
            </label>
            <textarea
              aria-invalid="false"
              class="oi-form-input form-control"
              id="query"
              name="query"
            />
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-2"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Save history query
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
